xarray/core/indexing.py

```diff
@@ -520,8 +532,17 @@ def shape(self) -> tuple[int, ...]:
         return tuple(shape)
 
     def __array__(self, dtype=None):
+        return np.asarray(self.get_duck_array(), dtype=dtype)
+
+    def get_duck_array(self):
         array = as_indexable(self.array)
-        return np.asarray(array[self.key], dtype=None)
+        array = array[self.key]
+        if isinstance(array, ExplicitlyIndexed):
+            array = array.get_duck_array()
+        if not is_duck_array(array):
+            # This hack is necessary for 0D netCDF4 variables
+            array = np.asarray(array)
```

Comment:
The zerodim test uses this dataset
```
expected = xr.Dataset()

expected["float_var"] = ([], 1.0e9, {"units": "units of awesome"})
expected["bytes_var"] = ([], b"foobar")
expected["string_var"] = ([], "foobar")
expected.to_netcdf("./zerodim.nc")
```

I think it would be cleaner (these 2 lines wouldn't be needed) if the NetCDF4ArrayWrapper always returned an array but it's not clear to me how to do that.

Reply:
Is there a reason why modifying the `_getitem` method on NetCDF4ArrayWrapper doesn't work? We could (and should) probably make it always return ndarray objects.